File: homework3/Python/Program1.py
from enum import Enum

# This initializes an enum class that defines three colors(RED, GREEN, BLUE) with associated integer values (1, 2, 3).
class Color(Enum):
	RED = 1
	GREEN = 2
	BLUE = 3



# This is the main function. Main function demonstrates the usage of enum where the color GREEN was chosen to display the name and value.
# The user can access the name by using chosen_name and the integer value by using chosen_value. 
# Since the enum only has 3 defined values, as soon as we try a value that is not defined in the enum Color, the program will raise a runtime error. 
# The reason why it's a runtime error is because Python has dynamic typing and does not enforce enum membership at compile time.
# The validation only occurs when Color(99) is actually executed, triggering ValueError in the enum constructor. 

def main():
	chosenColor = Color.GREEN
	print(f"Enum variable value: {chosenColor}")
	print(f"Enum variable name: {chosenColor.name}")
	print(f"Enum variable integer value: {chosenColor.value}")

	outOfRange = 99
	print(f"\nAttempting to create Color from integer {outOfRange}...")

	try:
		invalidColor = Color(outOfRange)
		print(f"Created enum: {invalidColor}")
	except ValueError as error:
		print("Python result: Runtime error (ValueError) when converting to enum.")
		print(f"Details: {error}")



main()

File: homework3/Csharp/Program1.cs
using System;

namespace Homework3.Csharp
{
    // Task 1: Enumeration Basics
    internal static class Program1
    {
        // C# enum can only be value-types such as int, long, and byte. By default int is used.
        private enum Color
        {
            Red = 1,
            Green = 2,
            Blue = 3
        }

        private static void Main()
        {
            Color chosenColor = Color.Green;

            Console.WriteLine($"Enum variable value: {chosenColor}");
            Console.WriteLine($"Enum variable name: {chosenColor}");
            Console.WriteLine($"Enum variable integer value: {(int)chosenColor}");

            int outOfRange = 99;
            Console.WriteLine();
            Console.WriteLine($"Attempting to cast integer {outOfRange} to Color...");

            // In C#, casting an out-of-range int to an enum does NOT throw at compile time
            // or at runtime. The enum simply holds an undefined underlying value.
            Color invalidColor = (Color)outOfRange;

            Console.WriteLine($"Resulting enum variable: {invalidColor} (numeric value {(int)invalidColor})");

            bool isDefined = Enum.IsDefined(invalidColor);
            Console.WriteLine($"Is {outOfRange} a defined Color value? {isDefined}");
        }
    }
}
